src/stateMachineFwk/stateMachineEngine.cpp

```diff
@@ -25,129 +25,156 @@
-
-		if(cb == NULL)
-			return;
-
-		if (cb->getControlBlockState() == FREE)
-		{
-		    log_msg(LOG_INFO, "Control block is freed\n");
-		    return;
-		}
-
-		while (true)
-		{
-			Event currentEvent;
-			if (cb->getCurrentEvent(currentEvent) == false)
-			{
-				break;
-			}
-
-			cmn::EventMessage * event_data = currentEvent.getEventData();
-
-			TempDataBlock *tempData = cb->getTempDataBlock();
-			if (tempData == NULL)
-			{
-			    log_msg(LOG_INFO, "Temp Data block is NULL"
-			            " for control block idx %d\n", cb->getCBIndex());
-
-			    if (event_data != NULL)
-			        delete event_data;
-
-			    break;
-			}
-
-			State *currentState_p = tempData->getCurrentState();
-			if (currentState_p == NULL)
-			{
-				log_msg(LOG_INFO, "Current state is NULL"
-						" for control block idx %d\n", cb->getCBIndex());
-
-				if (event_data != NULL)
-					delete event_data;
-
-				break;
-			}
-
-			if (currentState_p->validateEvent(*cb, tempData, currentEvent) == IGNORE)
-			{
-			    log_msg(LOG_INFO, "Event ignored for control block idx %d\n",
-			            cb->getCBIndex());
-
-			    if (event_data != NULL)
-			        delete event_data;
-
-			    continue;
-			}
-			ActStatus ret = handleProcedureEvent(*cb, *currentState_p, currentEvent);
-
-			if (ret == ABORT)
-			{
-				log_msg(LOG_INFO,"Abort Event Initiated \n");
-
-				Event abortEvent(ABORT_EVENT, NULL);
-				ret = handleProcedureEvent(*cb, *currentState_p, abortEvent);
-			}
-
-			if (event_data != NULL)
-			    delete event_data;
-
-			if (ret == HALT)
-				break;			
-
-		}
-
-		if (cb->getControlBlockState() == ALLOCATED)
-		    cb->setProcQueueFlag(false);
-	}
+StateMachineEngine::StateMachineEngine() :
+        procQ_m()
+{
+}
+
+StateMachineEngine::~StateMachineEngine()
+{
+}
+
+StateMachineEngine* StateMachineEngine::Instance()
+{
+    static StateMachineEngine SM;
+    return &SM;
 }
 
+bool StateMachineEngine::addCBToProcQ(ControlBlock *cb)
+{
+    return procQ_m.push(cb);
+}
 
+ActStatus StateMachineEngine::handleProcedureEvent(ControlBlock &cb,
+        StateMachineContext &smCtxt)
+{
+    State *currentState_p = smCtxt.tempDataBlock_mp->getCurrentState();
+
+    SmUtility *util = SmUtility::Instance();
+    log_msg(LOG_DEBUG,
+            "################ Executing actions for event: %s and State: %s #################\n",
+            util->convertEventToString(smCtxt.evt.getEventId()).c_str(),
+            util->convertStateToString(currentState_p->getStateId()).c_str());
+
+    time_t mytime = time(NULL);
+    debugEventInfo dEventInfo(smCtxt.evt.getEventId(),
+            currentState_p->getStateId(), mytime);
+    cb.addDebugInfo(dEventInfo);
+
+    return currentState_p->executeActions(smCtxt.evt.getEventId(), cb);
+}
+
+void StateMachineEngine::run()
+{
+
+    ControlBlock *cb = procQ_m.pop();
+
+    if (cb == NULL)
+        return;
+
+    if (cb->getControlBlockState() == FREE)
+    {
+        log_msg(LOG_INFO, "Control block is freed\n");
```

Comment:
can you keep assert ? or you think its not good idea ?

Reply:
This was just added as a safety check. Assert may not be a good idea. We can change the debug level to LOG_ERROR.